Introduce power-of-two FFFT

mrwojtek opened the pull request.

mrwojtek commented:
Still docstring to do but everything else is here.

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+class _TwiddleGate(cirq.ZPowGate):
```
Does this need to inherit from ZPowGate? It could just specify that gate as its decomposition.

mrwojtek replied:
Yea mean to inherit from `EigenGate`? I see that I would need to implement `__str__` and `__repr__` but I most likely should do that anyway. Except for `_eigen_components`, what am I require to support?

mrwojtek replied:
Thanks, done!

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+        return F0(*qubits)
+
+    if n % 2 != 0:
+        raise ValueError('Number of qubits is not a power of 2.')
```
This is not a correct check for powers of 2. It will e.g. allow the number 6 to pass.

You can use the expression `n & (n-1)` to detect non-powers-of-2. Note that this expression thinks zero is a power of two (because 2^{-infinity} = 0).

mrwojtek replied:
I was counting that this check will fail at the lower recursion levels, e.g. 6 will be split to ffft of 3, which will fail on parity check.

This seems much simpler than playing with the bits? Also, if #327 is merged (which is not decided yet), then this check is going away anyway.

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+    nx = n // ny
+    permutation = [(i % ny) * nx + (i // ny) for i in range(n)]
+
+    operations = list()
```
Nit: just `[]`

mrwojtek replied:
Done, but note that when I do it like that I get "This list creation could be rewritten as a list literal" PyCharm notice.

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+
+    ny = 2
+    nx = n // ny
+    permutation = [(i % ny) * nx + (i // ny) for i in range(n)]
```
```
def rotate_bits(v: int, n: int) -> int:
    return (v >> 1) | ((v & 1) << (n - 1))

permutation = [rotate_bits(i, n) for i in range(n)]
```

mrwojtek replied:
Well, I sent out #327 as well which handles generic composite case. If I follow your suggestion, it will not work later.

I can do as you are proposing and restore my old code back in #327? I can also just extract the magic of permutation initialization and just name it better?

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+
+    operations = list()
+
+    operations.append(_permute(qubits, permutation))
```
Store the permutation operation into a list, so that you can get the inverse via `cirq.inverse`.

mrwojtek replied:
That would be really cool, but how does it exactly works, both: cirq.inverse({0: 1, 1: 0, 2: 2, 3: 3}) and cirq.inverse([1, 0, 2, 3]) give "ZeroDivisionError: 0.0 cannot be raised to a negative power"?

mrwojtek replied:
That would be really cool, but how does it exactly works, both: cirq.inverse({0: 1, 1: 0, 2: 2, 3: 3}) and cirq.inverse([1, 0, 2, 3]) give "ZeroDivisionError: 0.0 cannot be raised to a negative power"?

kevinsung replied:
I believe that for @Strilanc's suggestion to work out, LinearPermutationGate must implement `__pow__` for -1, which it doesn't, but easily could.

kevinsung replied:
@mrwojtek now that LinearPermutationGate supports `cirq.inverse` you can resolve this, right?

mrwojtek replied:
Done!

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+    for y in range(ny):
+        operations.append(ffft(qubits[nx * y:nx * (y + 1)]))
+
+    operations.append(_permute(qubits, _inverse(permutation)))
```
`cirq.inverse` of the permutation operations

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+            operations.append(_TwiddleGate(x * y, n).on(qubits[ny * x + y]))
+        operations.append(ffft(qubits[ny * x:ny * (x + 1)]))
+
+    operations.append(_permute(qubits, permutation))
```
It re-permutes? That doesn't seem right.

mrwojtek replied:
Added documentation which explains that.

Strilanc commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'ω^%d_%d' % (self.k, self.n),
```
Prefer `'{}'.format(...)`

mrwojtek replied:
Done

Strilanc commented on openfermioncirq/primitives/_ffft_test.py:
```diff
@@ -0,0 +1,289 @@
+)
+
+
+def _state_from_amplitudes(amplitudes):
```
Type annotations. Better names.

I would call this something like a "single electron state" or similar.

mrwojtek replied:
Done.

Strilanc commented on openfermioncirq/primitives/_ffft_test.py:
```diff
@@ -0,0 +1,289 @@
+    return state
+
+
+def _fft_amplitudes(amplitudes):
```
We should also verify the effect on multi-electron states, since they will be moving around in an orthogonal space.

mrwojtek replied:
Done, note that helper method which creates multi-Fermion transforms is not trivial.

kevinsung commented on openfermioncirq/primitives/_ffft_test.py:
```diff
@@ -0,0 +1,289 @@
+        qubits_that_should_be_present=qubits
+    )
+
+    assert np.allclose(state, expected_state, rtol=0.0)
```
Add a test that creates two equivalent circuits, one using `ffft` and one using `bogoliubov_transform`, and checks that the circuits are equivalent.

mrwojtek replied:
Done.

kevinsung commented:
We've been omitting the `_` in file names, so remove the `_` prefix in the filenames.

kevinsung commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,196 @@
+
+    This gate assumes JWT representation of fermionic modes which are big-endian
+    encoded on consecutive qubits:
+    :math:`a_0^\dagger \lvert 0 \rangle = \lvert 10 \rangle` and
```
I'd say 00 instead of just 0.

mrwojtek replied:
Done

mrwojtek commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+F0 = _F0Gate()
+
+
+def ffft(qubits: Sequence[cirq.QubitId]) -> cirq.OP_TREE:
```
OK, I'll not add it now.

mrwojtek commented on openfermioncirq/primitives/_ffft.py:
```diff
@@ -0,0 +1,192 @@
+    This gate assumes JWT representation of fermionic modes which are big-endian
+    encoded on consecutive qubits:
+    :math:`a_0^\dagger \lvert 0 \rangle = \lvert 10_2 \rangle` and
+    :math:`a_1^\dagger \lvert 0 \rangle = \vert 01_2 \rangle`.
```
Done.

mrwojtek commented:
Alright, I finally added the docstring for this function, the PR is ready to final set of reviews!

I remember seeing a comment that I should rename this to _ffft. If I do that, I don't know how to test the internal helper functions `_inverse` and `_permute` because I don't know the way how to properly import this module.

I added tests for those helpers because in the beginning I had my own permutation circuit generator, which I removed after Bryan comment. If Craig's comment about inverting will start to work, I could actually remove all the internal helpers tests and just rename this file to ffft?

kevinsung commented:
I think we should just name the file ffft.py; I don't see a reason to add the underscore.

mrwojtek commented:
Kevin, I successfully renamed the file as you wanted, and dropped my own permutation inverse implementation. This PR is now complete and awaiting final approval.

kevinsung commented on openfermioncirq/primitives/ffft.py:
```diff
@@ -0,0 +1,261 @@
+        """Initializes Twiddle gate.
+
+        Args:
+            k: Nominator appearing in the exponent.
```
"Numerator" instead of "Nominator"
